### PR TITLE
feat: second-cut of using unenv to create a hybrid node.js compatibility setting (via inject)

### PR DESCRIPTION
**Pull request by IgorMinar**
rebased version of https://github.com/cloudflare/workers-sdk/pull/5220/files which uses unenv's `inject` to introduce globals

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included - the update to the fixture tests this
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [ ] I don't know
  - [ ] Required / Maybe required
  - [x] Not required because: no e2e tests are affected by node-compat
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): 
  - [x] Not necessary because: still experimental but we should follow up with a docs ticket.

### CODE REVIEW

**Comment by IgorMinar** [comment]
@petebacondarwin I rebased your PR but didn't want to force-push, so I'm creating a new PR. Let's chat tomorrow to discuss how we want to handle these PRs.

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (IgorMinar) [inline packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:24]: @petebacondarwin I don't understand what this resolve block is doing and why, could you please explain or suggest an inline comment?

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (IgorMinar) [inline packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:84]: todo: add global and globalThis assignment...

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:47]: Workers do not support:

```
const { resolve } = require("node:path");
```

We only support 

```
import { resolve } from "node:path";
```

In the first case you get:

```

✘ [ERROR] service core:user:worker-app: Uncaught Error: Dynamic require of "node:path" is not supported
```

So what this hack does is replace the dynamic require of the node with a load from a synthetic virtual file that contains:

```
export * from 'node:path';
```

Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:47]: I stole this from @dario-piotrowicz which he added to next-on-pages: https://sourcegraph.com/npm/cloudflare/next-on-pages@58cdf844a05934ade288ee4966a23bff1e983179/-/blob/dist/index.js

It is also used in:

- SST Remix: https://github.com/sst/ion/blob/4e1f667da8a555104d28efddd3a3e203566289f7/pkg/platform/src/components/cloudflare/remix.ts
- Open Next: https://github.com/sst/open-next/blob/65f7a6032d0dba552084becae50da9cf22145e09/packages/open-next/src/plugins/edge.ts
- Astro: https://github.com/withastro/astro/blob/8e465d621e7acc4fd36d78bef63450c6550e0ec1/packages/integrations/vercel/src/serverless/middleware.ts

Comment (IgorMinar) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:47]: Thanks for the explanation, Pete. I'll have to ponder this a bit, but I think it all makes sense.

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (petebacondarwin) [inline packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:85]:
```suggestion
			build.onResolve({ filter: UNENV_GLOBALS_RE }, ({ path }) => ({ path }));
```

Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:85]: Is the point of this resolve block to avoid esbuild trying to resolve the path itself, which it can't because these paths are virtual?

File: packages/wrangler/src/deploy/deploy.ts
Comment (IgorMinar) [reply packages/wrangler/src/deploy/deploy.ts:390]: hahahahahaha... i think it's duplicated 5 or 6 times. I was really itching to refactor this code but then I realized that each location has a slightly different way to emit errors and some other needs. I fear that refactoring this is a much bigger job than it seems, and I didn't want to expand the scope of this PR any more. I suggest we merge this as is, and then have someone on the wrangler team with the right knowledge refactor this to dedupe.

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (IgorMinar) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:33]: yeah. I think I asked the question first a while back, and then figure it out and left a comment. or something along those lines.

File: packages/wrangler/src/deploy/deploy.ts
Comment (IgorMinar) [reply packages/wrangler/src/deploy/deploy.ts:438]: you are right on both counts. 

perhaps we should drop this completely?

File: packages/wrangler/src/deploy/deploy.ts
Comment (IgorMinar) [reply packages/wrangler/src/deploy/deploy.ts:403]: what I wanted to do was to support the case when both flags are present in the compat flags list.

I just wanted to be nice and I also didn't want to add more code to error when both flags are present, since `nodejs_compat_v2` will be rolled into `nodejs_compat` eventually.

I don't feel very strongly about this though.

File: packages/wrangler/src/deployment-bundle/bundle.ts
Comment (IgorMinar) [reply packages/wrangler/src/deployment-bundle/bundle.ts:332]: yeah. I think I misunderstood the original intent. I'll defer to you on this one.

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (IgorMinar) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:9]: sgtm

File: packages/wrangler/src/deploy/deploy.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deploy/deploy.ts:390]: OK

File: packages/wrangler/src/deploy/deploy.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deploy/deploy.ts:403]: I understand that but this approach assumes the user implicitly understands the behaviour when both flags are enabled. It is kind of intuitive but...?

**Review by petebacondarwin — COMMENTED** [review]
Last small thing:

- a changeset (with a minor bump) - or are we claiming that since this is internal/experimental it should not appear in the changelog? I don't feel we should try to hide it by doing that: users can see the code being added so it is not a secret, and also no changeset means that the package will not be published.

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (IgorMinar) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:85]: yes

File: packages/wrangler/package.json
Comment (IgorMinar) [reply packages/wrangler/package.json:85]: yeah. but this won't happen until shortly before we drop the `experimental:` prefix.

**Review by RamIdeas — APPROVED** [review]
LGTM just needs the changeset before merge

File: packages/wrangler/src/deploy/deploy.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deploy/deploy.ts:438]: I tweaked the messaging.

File: packages/wrangler/src/deploy/deploy.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deploy/deploy.ts:403]: I'm going to make this a hard error for now. We can relax it if we find this is too restrictive and annoying.

File: fixtures/nodejs-hybrid-app/tests/index.test.ts
Comment (IgorMinar) [inline fixtures/nodejs-hybrid-app/tests/index.test.ts]: is this output stable enough not to break the CI unexpectidly in the future? maybe a smaller snapshot would be better (e.g. just check for presence of `{"id":"`

**Review by IgorMinar — COMMENTED** [review]
LGTM except for one regression that was introduced during the refactor. I didn't have enough time to debug it, but I discovered it in the matrix app and moved a reduction into the fixture as 0c2fc6d4fd19ab897c8313640cd63648fd1965e6

somehow don't tell esbuild that `node:assert` is a built-in, so it freaks out when it's not resolved to an absolute path. did we somehow forget to configure `externals`?

File: fixtures/nodejs-hybrid-app/tests/index.test.ts
Comment (petebacondarwin) [reply fixtures/nodejs-hybrid-app/tests/index.test.ts]: Indeed! The timestamp was non-deterministic.
I have changed it to only check for the id.

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (RamIdeas) [inline packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:145]: Does this work?

Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:145]: Our fixture doesn't test it but it is being used in the node.js compat matrix app that Igor and James C built. I believe that gives us enough confidence for today. But I'll see if I can add a quick check in the fixture.

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (RamIdeas) [inline packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:186]: @petebacondarwin can you check these too in the fixture test you proposed?

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (RamIdeas) [inline packages/wrangler/src/deployment-bundle/node-compat.ts:23]: 👏

File: packages/wrangler/src/dev.tsx
Comment (RamIdeas) [inline packages/wrangler/src/dev.tsx:576]: nit but these booleans could end up passed in the wrong order – better to make the params an object of named options

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:145]: Added 0ec2e8632ced68a2db402223a73f84b893643be7

File: packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts:186]: Done in 0ec2e8632ced68a2db402223a73f84b893643be7

File: packages/wrangler/src/dev.tsx
Comment (petebacondarwin) [reply packages/wrangler/src/dev.tsx:576]: Good call. I'll update.

File: packages/wrangler/src/pages/functions/buildPlugin.ts
Comment (RamIdeas) [inline packages/wrangler/src/pages/functions/buildPlugin.ts:51]: does this comment still apply?

Comment (petebacondarwin) [reply packages/wrangler/src/pages/functions/buildPlugin.ts:51]: Yes, basically what this does is set the nodejsCompat (aka v1) to true unless there is a mode provided. In this case there will never be a mode provided since there is still no way for the user to specify one.

Comment (petebacondarwin) [reply packages/wrangler/src/pages/functions/buildPlugin.ts:51]: I could have just hard-coded it to v1 but this way, when we do finally provide such a mechanism, this will pick it up straight away.

**Review by RamIdeas — APPROVED** [review]
You managed to add this extra compat flag and make the codebase cleaner 👏

File: packages/wrangler/src/pages/functions/buildPlugin.ts
Comment (RamIdeas) [reply packages/wrangler/src/pages/functions/buildPlugin.ts:51]: yeah that's what threw me off – I thought the mechanism had been provided since it tries using the var before defaulting to v1

**Review by RamIdeas — APPROVED** [review]

**Comment by petebacondarwin** [comment]
I'm going to override the GitGuardian check failure because it is picking up a DB secret that is actually a publicly accessible DB - so a false positive.

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/node-compat.ts:4]: Feel free to ignore this, but what about naming this with a bit more verbosity to describe what they do? Something like `legacy-build-time`/`runtime-only`/`hybrid-build-and-runtime`

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/node-compat.ts:45]: Given the flag is no longer experimental in workerd, do we want to keep this prefix stripping?

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (penalosa) [inline packages/wrangler/src/deployment-bundle/node-compat.ts:66]: This seems duplicative?

**Review by penalosa — APPROVED** [review]

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/node-compat.ts:4]: I see where you are coming from but I think the more verbose strings will actually make the code less readable.
I'll add a JSDOC comment to explain the values here.

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/node-compat.ts:45]: No we still want to require users of Wrangler to add this - it was orthogonal to workerd's experimental constraints - and is left here to avoid people trying to use this for production while we stabilize it.

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/node-compat.ts:66]: Agh! Cut and paste mistake. Good spot.

Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/node-compat.ts:66]: https://github.com/cloudflare/workers-sdk/pull/5991/commits/852db02ad3c8ff499c5910180c037b5a137f2f1d

File: packages/wrangler/src/deployment-bundle/node-compat.ts
Comment (petebacondarwin) [reply packages/wrangler/src/deployment-bundle/node-compat.ts:4]: https://github.com/cloudflare/workers-sdk/pull/5991/commits/852db02ad3c8ff499c5910180c037b5a137f2f1d

File: fixtures/nodejs-hybrid-app/wrangler.toml
Comment (petebacondarwin) [reply fixtures/nodejs-hybrid-app/wrangler.toml:11]: https://github.com/cloudflare/workers-sdk/pull/5991/commits/6298cd7199ecf1c5d760c3ba3fd34231e3c158db

File: fixtures/nodejs-hybrid-app/src/index.ts
Comment (petebacondarwin) [reply fixtures/nodejs-hybrid-app/src/index.ts:25]: https://github.com/cloudflare/workers-sdk/pull/5991/commits/91d257651513d23b1b12f82d3b4fd412772b0c03

**Pull request merged**
